File: calculation-engine/src/ziwei.js
import iztro from 'iztro';
import { evidence, blend, moduleResult, clamp, mod, boundedCache } from './core.js';
import { hourBranch } from './time.js';

const CONFIG={yearDivide:'normal',horoscopeDivide:'normal',ageDivide:'normal',dayDivide:'current',algorithm:'default'};
const MAJOR={'紫微':[.2,-.2],'天机':[.1,.5],'太阳':[.3,.2],'武曲':[.25,-.2],'天同':[-.1,-.35],'廉贞':[.1,.25],'天府':[.2,-.5],'太阴':[-.1,-.25],'贪狼':[.15,.5],'巨门':[-.15,.2],'天相':[.15,-.4],'天梁':[.1,-.45],'七杀':[.1,.55],'破军':[-.1,.65]};
const AUX={'左辅':[.2,-.1],'右弼':[.2,-.1],'文昌':[.2,0],'文曲':[.2,0],'天魁':[.2,-.1],'天钺':[.2,-.1],'擎羊':[-.25,.2],'陀罗':[-.25,-.1],'火星':[-.25,.2],'铃星':[-.25,.15],'地空':[-.25,.3],'地劫':[-.25,.3]};
const BRIGHTNESS={'庙':1,'旺':.95,'得':.85,'利':.8,'平':.7,'不':.6,'陷':.5};
const TRANSFORMS={'禄':[.35,-.1],'权':[.2,.2],'科':[.25,-.2],'忌':[-.45,.15]};
const LAYERS={natal:.15,decadal:.15,yearly:.15,monthly:.10,daily:.20,hourly:.25};
const proximity=(p,t)=>({0:1,4:.6,8:.6,6:.5}[mod(p-t,12)]??0);

// Category selects which palace(s) the reading is measured against, replacing a
// single hard-coded 命宫 target. Provider alias sets are listed together.
// Symbolic editorial emphases, not a claim about another person.
const target=(palaces,weight)=>Object.freeze({palaces:Object.freeze(palaces),weight});
export const ZIWEI_CATEGORY_TARGETS=Object.freeze({
  general:Object.freeze([target(['命宫'],1)]),
  other:Object.freeze([target(['命宫'],1)]),
  love:Object.freeze([target(['夫妻'],.75),target(['福德'],.25)]),
  career:Object.freeze([target(['官禄'],.80),target(['迁移'],.20)]),
  money:Object.freeze([target(['财帛'],.75),target(['田宅'],.25)]),
  study:Object.freeze([target(['官禄'],.45),target(['父母'],.30),target(['福德'],.25)]),
  friends:Object.freeze([target(['仆役','交友'],.65),target(['兄弟'],.35)]),
});

for (const [category,spec] of Object.entries(ZIWEI_CATEGORY_TARGETS)) {
  if(!spec.length)throw new Error(`ZIWEI_TARGETS_EMPTY:${category}`);
  if(!spec.every(t=>t.palaces.length&&Number.isFinite(t.weight)&&t.weight>0))throw new Error(`ZIWEI_TARGET_VALUE:${category}`);
  if(Math.abs(spec.reduce((s,t)=>s+t.weight,0)-1)>1e-9)throw new Error(`ZIWEI_TARGETS_NOT_NORMALIZED:${category}`);
}

/** Resolves each target to a palace index, failing loudly when a palace is absent. */
function resolveTargets(chart,category) {
  if(!Object.hasOwn(ZIWEI_CATEGORY_TARGETS,category))throw new Error('INVALID_CATEGORY');
  return ZIWEI_CATEGORY_TARGETS[category].map(({palaces,weight})=>{
    const index=chart.palaces.findIndex(p=>palaces.includes(p.name));
    if(index<0)throw new Error(`ZIWEI_PALACE_UNRESOLVED:${palaces.join('/')}`);
    return {palace:chart.palaces[index].name,aliases:palaces,index,weight};
  });
}

const weightedProximity=(palaceIndex,targets)=>targets.reduce((sum,t)=>sum+t.weight*proximity(palaceIndex,t.index),0);
const charts=boundedCache(256);
function configure(){iztro.astro.config(CONFIG);iztro.i18n?.setLanguage?.('zh-CN');}
function chartFor(date,index,gender) {
  const key=`${date}|${index}|${gender}`;
  if(charts.get(key))return charts.get(key);
  configure();
  return charts.set(key,iztro.astro.bySolar(date,index,gender==='male'?'男':'女',true,'zh-CN'));
}
export function buildZiWei(birth,traditionalProfile) {
  const indexes=birth.clock?[hourBranch(Number(birth.clock.slice(0,2)))]:Array.from({length:12},(_,i)=>i);
  if(birth.status==='birth_time_nonexistent')return {charts:[],reason:birth.status,unknownHour:!birth.clock};
  // Unknown convention is modeled as alternatives, never inferred from a name.
  const genders=['male','female'].includes(traditionalProfile)?[traditionalProfile]:['male','female'];
  const entries=indexes.flatMap(index=>genders.map(gender=>({index,gender,chart:chartFor(birth.date,index,gender)})));
  return {charts:entries,unknownHour:!birth.clock,unknownConvention:genders.length>1,rules:CONFIG};
}
function scoreChart(chart,date,timeIndex,category) {
  configure();
  const targets=resolveTargets(chart,category);
  const stars=chart.palaces.flatMap((p,index)=>[...p.majorStars,...p.minorStars,...p.adjectiveStars].map(s=>({...s,palace:index})));
  const parts=[];
  for(const [weight,catalog,divisor]of [[.45,MAJOR,4],[.20,AUX,3]]) {
    const selected=stars.filter(s=>s.name in catalog);
    if(selected.length!==Object.keys(catalog).length||new Set(selected.map(s=>s.name)).size!==selected.length)throw new Error('ZIWEI_STAR_CATALOG_MISMATCH');
    const values=[0,1].map(k=>selected.reduce((sum,s)=>{
      if(catalog===MAJOR&&!(s.brightness in BRIGHTNESS))throw new Error('ZIWEI_MAJOR_BRIGHTNESS_MISSING');
      return sum+catalog[s.name][k]*(BRIGHTNESS[s.brightness]??1)*weightedProximity(s.palace,targets)/divisor;
    },0));
    parts.push([weight,evidence(...values.map(clamp),1)]);
  }
  const transformed=[],layerDetails={};
  const natalEvents=stars.filter(s=>s.mutagen in TRANSFORMS).map(s=>({name:s.name,kind:s.mutagen,palace:s.palace}));
  const horoscope=chart.horoscope(date,timeIndex);
  for(const [layer,weight]of Object.entries(LAYERS)) {
    const data=horoscope[layer], index=layer==='natal'?null:data.index;
    if(index!==null&&index<0)continue;
    const events=layer==='natal'?natalEvents:data.mutagen.map((name,i)=>({name,kind:['禄','权','科','忌'][i],palace:stars.find(s=>s.name===name)?.palace}));
    if(events.length!==4||events.some(e=>e.palace==null)||new Set(events.map(e=>e.kind)).size!==4)throw new Error('ZIWEI_TRANSFORM_CATALOG_MISMATCH');
    // The natal layer is measured against the category's target palaces; the
    // time layers keep their own horoscope palace.
    const reach=e=>index===null?weightedProximity(e.palace,targets):proximity(e.palace,index);
    const v=[0,1].map(k=>clamp(events.reduce((sum,e)=>sum+reach(e)*TRANSFORMS[e.kind][k]/2,0)));
    transformed.push([weight,evidence(...v,1)]);
    layerDetails[layer]={target:index===null?targets.map(t=>({palace:t.palace,index:t.index,weight:t.weight})):index,events};
  }
  parts.push([.35,blend(transformed)]);
  return {evidence:blend(parts),layers:layerDetails,category,
    targetPalaces:targets.map(t=>({palace:t.palace,aliases:t.aliases,index:t.index,weight:t.weight})),
    lifePalace:chart.palaces.findIndex(p=>p.name==='命宫'),bodyPalace:chart.palaces.findIndex(p=>p.isBodyPalace),
    majorCount:14,auxiliaryCount:12};
}
export function scoreZiWei(built,cal,category='general') {
  if(!built.charts.length)return moduleResult(evidence(),{reason:built.reason,category});
  const results=built.charts.map(entry=>({...scoreChart(entry.chart,cal.local.date,cal.hour.branch,category),birthHourIndex:entry.index,convention:entry.gender}));
  if(results.length===1)return moduleResult(results[0].evidence,{...results[0],rules:CONFIG,scenarioCount:1,method:'single_chart'});
  const ranges=Object.fromEntries(['a','c'].map(k=>[k,{min:Math.min(...results.map(r=>r.evidence[k])),max:Math.max(...results.map(r=>r.evidence[k]))}]));
  const conservative=k=>ranges[k].min>0?ranges[k].min:ranges[k].max<0?ranges[k].max:0;
  // Missing birth inputs and cross-scenario agreement are separate quantities.
  // The cap is explicit product policy, NOT a likelihood of any birth hour.
  const inputFactor=(built.unknownHour?.75:1)*(built.unknownConvention?.75:1);
  const q=Math.min(...results.map(r=>r.evidence.coverage))*inputFactor;
  return moduleResult(evidence(conservative('a'),conservative('c'),q),{method:'conservative_across_possible_charts',scenarioCount:results.length,
    unknownBirthHour:built.unknownHour,unknownConvention:built.unknownConvention,inputCoverageFactor:inputFactor,
    category,targetPalaces:results[0].targetPalaces,
    ranges,scenarioScores:results.map(({birthHourIndex,convention,evidence:e})=>({birthHourIndex,convention,...e})),rules:CONFIG},'scenario_analysis');
}
export function inspectZiWei(built) {
  return built.charts.map(({index,gender,chart})=>({birthHourIndex:index,convention:gender,
    // Provider chart is available for audit, but not needed by the app result.
    chart:JSON.parse(JSON.stringify(chart))}));
}

File: mobile_app/lib/local_engine/ziwei/ziwei.dart
/// Port of `calculation-engine/src/ziwei.js`.
///
/// Scores the natal chart and five time layers against the palaces the chosen
/// category cares about. When the birth hour or the traditional convention is
/// unknown, every possible chart is built and the *conservative* agreement
/// across them is taken, with an explicit coverage penalty — the engine never
/// picks a likely hour or infers a convention from a name.
library;

import '../core/bounded_cache.dart';
import '../core/core.dart';
import '../core/numbers.dart';
import '../calendar/calendar.dart';
import '../time/local_time.dart' show BirthContext, hourBranch;
import 'ziwei_chart.dart';

const Map<String, List<double>> _majorStars = <String, List<double>>{
  '紫微': <double>[.2, -.2],
  '天机': <double>[.1, .5],
  '太阳': <double>[.3, .2],
  '武曲': <double>[.25, -.2],
  '天同': <double>[-.1, -.35],
  '廉贞': <double>[.1, .25],
  '天府': <double>[.2, -.5],
  '太阴': <double>[-.1, -.25],
  '贪狼': <double>[.15, .5],
  '巨门': <double>[-.15, .2],
  '天相': <double>[.15, -.4],
  '天梁': <double>[.1, -.45],
  '七杀': <double>[.1, .55],
  '破军': <double>[-.1, .65],
};

const Map<String, List<double>> _auxiliaryStars = <String, List<double>>{
  '左辅': <double>[.2, -.1],
  '右弼': <double>[.2, -.1],
  '文昌': <double>[.2, 0],
  '文曲': <double>[.2, 0],
  '天魁': <double>[.2, -.1],
  '天钺': <double>[.2, -.1],
  '擎羊': <double>[-.25, .2],
  '陀罗': <double>[-.25, -.1],
  '火星': <double>[-.25, .2],
  '铃星': <double>[-.25, .15],
  '地空': <double>[-.25, .3],
  '地劫': <double>[-.25, .3],
};

const Map<String, double> _brightnessWeights = <String, double>{
  '庙': 1,
  '旺': .95,
  '得': .85,
  '利': .8,
  '平': .7,
  '不': .6,
  '陷': .5,
};

const Map<String, List<double>> _transforms = <String, List<double>>{
  '禄': <double>[.35, -.1],
  '权': <double>[.2, .2],
  '科': <double>[.25, -.2],
  '忌': <double>[-.45, .15],
};

const Map<String, double> _layers = <String, double>{
  'natal': .15,
  'decadal': .15,
  'yearly': .15,
  'monthly': .10,
  'daily': .20,
  'hourly': .25,
};

double _proximity(int palace, int target) {
  final offset = jsMod(palace - target, 12);
  return switch (offset) {
    0 => 1,
    4 => .6,
    8 => .6,
    6 => .5,
    _ => 0,
  };
}

/// One category target: the palace names it accepts (provider aliases listed
/// together) and how much of the reading it carries.
class ZiweiTarget {
  const ZiweiTarget(this.palaces, this.weight);

  final List<String> palaces;
  final double weight;
}

/// Category selects which palace(s) the reading is measured against, replacing
/// a single hard-coded 命宫 target. Symbolic editorial emphases, not a claim
/// about another person.
const Map<String, List<ZiweiTarget>> ziweiCategoryTargets =
    <String, List<ZiweiTarget>>{
      'general': <ZiweiTarget>[
        ZiweiTarget(<String>['命宫'], 1),
      ],
      'other': <ZiweiTarget>[
        ZiweiTarget(<String>['命宫'], 1),
      ],
      'love': <ZiweiTarget>[
        ZiweiTarget(<String>['夫妻'], .75),
        ZiweiTarget(<String>['福德'], .25),
      ],
      'career': <ZiweiTarget>[
        ZiweiTarget(<String>['官禄'], .80),
        ZiweiTarget(<String>['迁移'], .20),
      ],
      'money': <ZiweiTarget>[
        ZiweiTarget(<String>['财帛'], .75),
        ZiweiTarget(<String>['田宅'], .25),
      ],
      'study': <ZiweiTarget>[
        ZiweiTarget(<String>['官禄'], .45),
        ZiweiTarget(<String>['父母'], .30),
        ZiweiTarget(<String>['福德'], .25),
      ],
      'friends': <ZiweiTarget>[
        ZiweiTarget(<String>['仆役', '交友'], .65),
        ZiweiTarget(<String>['兄弟'], .35),
      ],
    };

/// Validated at startup, as the Node engine validates at module load.
void validateZiweiTargets() {
  for (final entry in ziweiCategoryTargets.entries) {
    final spec = entry.value;
    if (spec.isEmpty) throw EngineFailure('ZIWEI_TARGETS_EMPTY:${entry.key}');
    if (!spec.every(
      (t) => t.palaces.isNotEmpty && t.weight.isFinite && t.weight > 0,
    )) {
      throw EngineFailure('ZIWEI_TARGET_VALUE:${entry.key}');
    }
    final total = spec.fold<double>(0, (s, t) => s + t.weight);
    if ((total - 1).abs() > 1e-9) {
      throw EngineFailure('ZIWEI_TARGETS_NOT_NORMALIZED:${entry.key}');
    }
  }
}

class _ResolvedTarget {
  const _ResolvedTarget(this.palace, this.aliases, this.index, this.weight);

  final String palace;
  final List<String> aliases;
  final int index;
  final double weight;
}

/// Resolves each target to a palace index, failing loudly when a palace is
/// absent rather than silently scoring against the wrong one.
List<_ResolvedTarget> _resolveTargets(ZiweiAstrolabe chart, String category) {
  final spec = ziweiCategoryTargets[category];
  if (spec == null) throw const EngineFailure('INVALID_CATEGORY');
  return <_ResolvedTarget>[
    for (final target in spec)
      () {
        final index = chart.palaces.indexWhere(
          (p) => target.palaces.contains(p.name),
        );
        if (index < 0) {
          throw EngineFailure(
            'ZIWEI_PALACE_UNRESOLVED:${target.palaces.join('/')}',
          );
        }
        return _ResolvedTarget(
          chart.palaces[index].name,
          target.palaces,
          index,
          target.weight,
        );
      }(),
  ];
}

double _weightedProximity(int palaceIndex, List<_ResolvedTarget> targets) =>
    targets.fold<double>(
      0,
      (sum, t) => sum + t.weight * _proximity(palaceIndex, t.index),
    );

final BoundedCache<String, ZiweiAstrolabe> _chartCache =
    BoundedCache<String, ZiweiAstrolabe>(256);

ZiweiAstrolabe _chartFor(String date, int index, String gender) {
  final key = '$date|$index|$gender';
  final cached = _chartCache.get(key);
  if (cached != null) return cached;
  return _chartCache.set(
    key,
    buildAstrolabe(solarDate: date, timeIndex: index, gender: gender),
  );
}

/// Every chart that could describe this birth, given what is known.
class BuiltZiWei {
  const BuiltZiWei({
    required this.charts,
    required this.unknownHour,
    required this.unknownConvention,
    this.reason,
  });

  final List<({int index, String gender, ZiweiAstrolabe chart})> charts;
  final bool unknownHour;
  final bool unknownConvention;
  final String? reason;
}

BuiltZiWei buildZiWei(BirthContext birth, String? traditionalProfile) {
  final clock = birth.clock;
  if (birth.status == 'birth_time_nonexistent') {
    return BuiltZiWei(
      charts: const <({int index, String gender, ZiweiAstrolabe chart})>[],
      unknownHour: clock == null,
      unknownConvention: false,
      reason: birth.status,
    );
  }
  final indexes = clock != null
      ? <int>[hourBranch(int.parse(clock.substring(0, 2)))]
      : <int>[for (var i = 0; i < 12; i++) i];
  // An unknown convention is modelled as alternatives, never inferred from a
  // name.
  final genders =
      (traditionalProfile == 'male' || traditionalProfile == 'female')
      ? <String>[traditionalProfile!]
      : <String>['male', 'female'];

  return BuiltZiWei(
    charts: <({int index, String gender, ZiweiAstrolabe chart})>[
      for (final index in indexes)
        for (final gender in genders)
          (
            index: index,
            gender: gender,
            chart: _chartFor(birth.date, index, gender),
          ),
    ],
    unknownHour: clock == null,
    unknownConvention: genders.length > 1,
  );
}

class _ChartScore {
  const _ChartScore(this.evidence, this.targets);

  final Evidence evidence;
  final List<_ResolvedTarget> targets;
}

_ChartScore _scoreChart(
  ZiweiAstrolabe chart,
  String date,
  int timeIndex,
  String category,
) {
  final targets = _resolveTargets(chart, category);
  final stars = chart.stars;
  final parts = <(double, Evidence)>[];

  for (final spec in <(double, Map<String, List<double>>, double, bool)>[
    (.45, _majorStars, 4, true),
    (.20, _auxiliaryStars, 3, false),
  ]) {
    final weight = spec.$1;
    final catalog = spec.$2;
    final divisor = spec.$3;
    final requireBrightness = spec.$4;
    final selected = stars.where((s) => catalog.containsKey(s.name)).toList();
    if (selected.length != catalog.length ||
        selected.map((s) => s.name).toSet().length != selected.length) {
      throw const EngineFailure('ZIWEI_STAR_CATALOG_MISMATCH');
    }
    final values = <double>[
      for (var k = 0; k < 2; k++)
        selected.fold<double>(0, (sum, s) {
          if (requireBrightness &&
              !_brightnessWeights.containsKey(s.brightness)) {
            throw const EngineFailure('ZIWEI_MAJOR_BRIGHTNESS_MISSING');
          }
          return sum +
              catalog[s.name]![k] *
                  (_brightnessWeights[s.brightness] ?? 1) *
                  _weightedProximity(s.palace, targets) /
                  divisor;
        }),
    ];
    parts.add((
      weight,
      Evidence(clampUnit(values[0]), clampUnit(values[1]), 1),
    ));
  }

  final transformed = <(double, Evidence)>[];
  final natalEvents = <({String kind, int palace})>[
    for (final s in stars)
      if (_transforms.containsKey(s.mutagen))
        (kind: s.mutagen, palace: s.palace),
  ];
  final horoscope = horoscopeFor(chart, date, timeIndex);

  for (final entry in _layers.entries) {
    final layer = entry.key;
    final data = horoscope[layer];
    final index = layer == 'natal' ? null : data!.index;
    if (index != null && index < 0) continue;

    final events = layer == 'natal'
        ? natalEvents
        : <({String kind, int palace})>[
            for (var i = 0; i < data!.mutagen.length; i++)
              (
                kind: _mutagenKinds[i],
                palace: _palaceOfStar(stars, data.mutagen[i]),
              ),
          ];
    if (events.length != 4 ||
        events.any((e) => e.palace < 0) ||
        events.map((e) => e.kind).toSet().length != 4) {
      throw const EngineFailure('ZIWEI_TRANSFORM_CATALOG_MISMATCH');
    }

    // The natal layer is measured against the category's target palaces; the
    // time layers keep their own horoscope palace.
    double reach(({String kind, int palace}) e) => index == null
        ? _weightedProximity(e.palace, targets)
        : _proximity(e.palace, index);

    final v = <double>[
      for (var k = 0; k < 2; k++)
        clampUnit(
          events.fold<double>(
            0,
            (sum, e) => sum + reach(e) * _transforms[e.kind]![k] / 2,
          ),
        ),
    ];
    transformed.add((entry.value, Evidence(v[0], v[1], 1)));
  }

  parts.add((.35, blend(transformed)));
  return _ChartScore(blend(parts), targets);
}

/// The four transformation kinds, in 禄权科忌 order.
const List<String> _mutagenKinds = <String>['禄', '权', '科', '忌'];

/// The palace a named star sits in, or -1 when the chart does not carry it.
int _palaceOfStar(List<ZiweiStar> stars, String name) {
  for (final star in stars) {
    if (star.name == name) return star.palace;
  }
  return -1;
}

ModuleResult scoreZiWei(
  BuiltZiWei built,
  CalendarAt cal, [
  String category = 'general',
]) {
  if (built.charts.isEmpty) {
    return ModuleResult(Evidence(), <String, Object?>{
      'reason': built.reason,
      'category': category,
    });
  }

  final results =
      <({Evidence evidence, int birthHourIndex, String convention})>[
        for (final entry in built.charts)
          (
            evidence: _scoreChart(
              entry.chart,
              cal.local.date,
              cal.hour.branch,
              category,
            ).evidence,
            birthHourIndex: entry.index,
            convention: entry.gender,
          ),
      ];

  if (results.length == 1) {
    return ModuleResult(results[0].evidence, <String, Object?>{
      'category': category,
      'scenarioCount': 1,
      'method': 'single_chart',
    });
  }

  double minOf(double Function(Evidence) pick) =>
      results.map((r) => pick(r.evidence)).reduce((a, b) => a < b ? a : b);
  double maxOf(double Function(Evidence) pick) =>
      results.map((r) => pick(r.evidence)).reduce((a, b) => a > b ? a : b);

  double conservative(double Function(Evidence) pick) {
    final low = minOf(pick);
    final high = maxOf(pick);
    if (low > 0) return low;
    if (high < 0) return high;
    return 0;
  }

  // Missing birth inputs and cross-scenario agreement are separate quantities.
  // The cap is explicit product policy, NOT a likelihood of any birth hour.
  final inputFactor =
      (built.unknownHour ? .75 : 1) * (built.unknownConvention ? .75 : 1);
  final q = minOf((e) => e.coverage) * inputFactor;

  return ModuleResult(
    Evidence(conservative((e) => e.a), conservative((e) => e.c), q),
    <String, Object?>{
      'method': 'conservative_across_possible_charts',
      'scenarioCount': results.length,
      'unknownBirthHour': built.unknownHour,
      'unknownConvention': built.unknownConvention,
      'inputCoverageFactor': inputFactor,
      'category': category,
    },
    'scenario_analysis',
  );
}
